prime-dotnet-webapi/Services/CommunitySiteService.cs

```diff
@@ -196,40 +196,41 @@ private void UpdateBusinessHours(Site current, CommunitySiteUpdateModel updated)
         {
-            // Wholesale replace the remote users
-            foreach (var remoteUser in current.RemoteUsers)
+            if (updated.RemoteUsers == null)
             {
-                foreach (var certification in remoteUser.RemoteUserCertifications)
-                {
-                    _context.Remove(certification);
-                }
-                _context.RemoteUsers.Remove(remoteUser);
+                return;
             }
 
-            if (updated?.RemoteUsers != null && updated?.RemoteUsers.Count() != 0)
+            foreach (var updateRemoteUser in updated.RemoteUsers)
             {
-                foreach (var remoteUser in updated.RemoteUsers)
-                {
-                    remoteUser.SiteId = current.Id;
-                    var remoteUserCertifications = new List<RemoteUserCertification>();
+                var existingRemoteUser = _context.RemoteUsers
+                    .FirstOrDefault(u => u.Id == updateRemoteUser.Id);
 
-                    foreach (var certification in remoteUser.RemoteUserCertifications)
+                if (existingRemoteUser == null)
+                {
+                    current.RemoteUsers.Add(updateRemoteUser);
+                }
+                else
+                {
+                    _context.Entry(existingRemoteUser).CurrentValues.SetValues(updateRemoteUser);
+                    foreach (var certification in existingRemoteUser.RemoteUserCertifications)
                     {
-                        var newCertification = new RemoteUserCertification
+                        var existingCertification = existingRemoteUser.RemoteUserCertifications
+                            .FirstOrDefault(c => c.Id == certification.Id);
+
+                        if (existingCertification == null)
                         {
-                            RemoteUser = remoteUser,
-                            CollegeCode = certification.CollegeCode,
-                            LicenseNumber = certification.LicenseNumber,
-                            LicenseCode = certification.LicenseCode
-                        };
-                        _context.Entry(newCertification).State = EntityState.Added;
-                        remoteUserCertifications.Add(newCertification);
+                            existingRemoteUser.RemoteUserCertifications.Add(certification);
+                        }
+                        else
+                        {
+                            _context.Entry(existingCertification).CurrentValues.SetValues(certification);
+                        }
                     }
-                    remoteUser.RemoteUserCertifications = remoteUserCertifications;
-
-                    _context.Entry(remoteUser).State = EntityState.Added;
                 }
             }
+
+            _context.RemoteUsers.RemoveRange(current.RemoteUsers.Where(u => !updated.RemoteUsers.Contains(u)));
```

Comment:
Calling `.Contains` on a collection of objects isn't going to work the way you hope, unfortunately. 
![image](https://user-images.githubusercontent.com/39168456/140586601-741c9904-d5e7-4a32-9995-90f24d67c6f2.png)

The default equality comparer for objects is reference equality, i.e, are they the same object. This will be false for every member since they are not the same object (one set of objects came from the DB and the other came from the FE).

As an aside, you _could_ implement the `.Equals()` on `RemoteUser` to have a custom comparer, but there are several problems with that. For one, you would also have to implement `.GetHashCode()` and this is kinda polluting the data model at that point. The other issue is that it isn't really the correct way to do it anyway. I would not expect `.Equals` on two objects to only test the Id to say they are the same object. These are not the same object:

```C#
var person1 = new Person
{
    Id = 1,
    Name = "Alice"
};

var person2 = new Person
{
    Id = 1,
    Name = "Bob"
};
```

so I would not expect them to return true if I did `person1.Equals(person2)`.  In reality they may represent the same *database entry*, witch is not the same concept and so is not accurately handled by `.Equals()` or `.Contains()`.

Reply:
In the foreach loop, new remote users are added by `current.RemoteUsers.Add()`, and existing remote user are updated thru `_context.Entry().SetValues()`. Somehow using `Contains` here does work magically (tested) likely due to EF uses reference in SetValues() maybe?

Reply:
Done some tests, and it seems checking equality by reference is valid here.

Reply:
Due to this line of code, existing `RemoteUser`s are being dropped and re-added rather than updated.

Reply:
Please see the failing unit test I added

Reply:
Thanks for the unit tests! Pushed more fix.